examples: add gatekeeper-auth example. (#1963)
electric-sql/electric: examples/gatekeeper-auth/README.md

# Electric Gatekeeper Auth Example

This example demonstrates a number of ways of implementing the [Gatekeeper auth pattern](https://electric-sql.com/docs/guides/auth#gatekeeper) for securing access to the [Electric sync service](https://electric-sql.com/product/sync).

It includes:

- an [`./api`](./api) service for generating auth tokens
- three options for validating those auth tokens when proxying requests to Electric:
    - [`./api`](./api) the API itself
    - [`./caddy`](./caddy) a Caddy web server as a reverse proxy
    - [`./edge`](./edge) an edge function that you can run in front of a CDN

## How it works

There are two steps to the gatekeeper pattern:

1. first a client posts authentication credentials to a gatekeeper endpoint to generate an auth token
2. the client then makes requests to Electric via an authorising proxy that validates the auth token against the shape request

> The auth token can be *shape-scoped* (i.e.: can sign a specific shape definition). This allows the proxy to authorise a shape request by comparing the shape signed into the token with the [shape defined in the request parameters](https://electric-sql.com/docs/quickstart#http-api). This allows you to:

Reviewer: Nit: i found the explanation of signing the shape in the token a bit confusing. I think it would be more accurate and clear to use the "claim" terminology. This is what i propose:

> 
```suggestion
The auth token can be *shape-scoped* (i.e.: include a claim containing the shape definition). This allows the proxy to authorise a shape request by comparing the shape claim from the token with the [shape defined in the request parameters](https://electric-sql.com/docs/quickstart#http-api). This allows you to:
```

- keep your main authorisation logic in your API (in the gatekeeper endpoint) where it's natural to do things like query the database and call external authorisation services; and to
- run your authorisation logic *once* when generating a token, rather than on the "hot path" of every shape request in your authorising proxy

### Implementation

The core of this example is an [Elixir/Phoenix](https://www.phoenixframework.org) web application in [`./api`](./api). This exposes (in [`api_web/router.ex`](./api/lib/api_web/router.ex)):

1. a gatekeeper endpoint at `POST /gatekeeper/:table`
2. a proxy endpoint at `GET /proxy/v1/shape`

<a href="./docs/img/gatekeeper-flow.jpg" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="./docs/img/gatekeeper-flow.dark.png" />
    <source media="(prefers-color-scheme: light)" srcset="./docs/img/gatekeeper-flow.png" />
    <img alt="Gatekeeper flow diagramme" src="./docs/img/gatekeeper-flow.jpg"
        width="100%"
    />
  </picture>

> </a>

Reviewer: Nit: in this diagram it looks like Electric replies directly to the client but actually it replies to the proxy which forwards it to the client. Perhaps we want to make this explicit in this diagram as it may otherwise cause confusion with the troubleshooting guide which will explain that the proxy must not strip the electric headers on the response.

EDIT: just noted that the explanation explain this. I'll let you decide whether the diagram needs to show this more explicitly or not.

```

In a new terminal, make a `POST` request to the gatekeeper endpoint:

```console
$ curl -sX POST "http://localhost:4000/gatekeeper/items" | jq
{
  "headers": {
    "Authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJKb2tlbiIsImV4cCI6MTczMTUwMjM2OSwiaWF0IjoxNzMxNDk1MTY5LCJpc3MiOiJKb2tlbiIsImp0aSI6IjMwM28zYmx0czN2aHYydXNiazAwMDJrMiIsIm5iZiI6MTczMTQ5NTE2OSwic2hhcGUiOnsibmFtZXNwYWNlIjpudWxsLCJ0YWJsZSI6Iml0ZW1zIiwid2hlcmUiOm51bGwsImNvbHVtbnMiOm51bGx9fQ.8UZehIWk1EDQ3dJ4ggCBNkx9vGudfrD9appqs8r6zRI"
  },
  "url": "http://localhost:3000/proxy/v1/shape",
  "table": "items"
}
```

You'll see that the response contains:

- the proxy `url` to make shape requests to (`http://localhost:4000/proxy/v1/shape`)
- the request parameters for the shape we're requesting, in this case `"table": "items"`
- an `Authorization` header, containing a `Bearer <token>`

Copy the auth token and set it to an env var:

```shell
export AUTH_TOKEN="<token>"
```

First let's make a `GET` request to the proxy endpoint *without* the auth token. It will be rejected with a `403` status:

```console
$ curl -sv "http://localhost:4000/proxy/v1/shape?table=items&offset=-1"
...

> < HTTP/1.1 403 Forbidden

Reviewer: Shouldn't the proxy return a 401 as in the Caddy example?

...
```

Now let's add the authorization header. The request will be successfully proxied through to Electric:

```console
$ curl -sv --header "Authorization: Bearer ${AUTH_TOKEN}" \
      "http://localhost:4000/proxy/v1/shape?table=items&offset=-1"
...
< HTTP/1.1 200 OK
...
```

However if we try to request a different shape (i.e.: using different request parameters), the request will not match the shape signed into the auth token claims and will be rejected:

```console
$ curl -sv --header "Authorization: Bearer ${AUTH_TOKEN}" \
      "http://localhost:4000/proxy/v1/shape?table=items&offset=-1&where=true"
...
< HTTP/1.1 403 Forbidden
...
```

Note that we got an empty response when successfully proxied through to Electric above because there are no `items` in the database. If you like, you can create some, e.g. using `psql`:

```console

> $ psql "postgresql://postgres:password@localhost:54321/electric?sslmode=disable"

Reviewer: We're using `psql` here but it's not listed in the prerequisites.

psql (16.4)
Type "help" for help.

electric=# \d
               List of relations
 Schema |       Name        | Type  |  Owner
--------+-------------------+-------+----------
 public | items             | table | postgres
 public | schema_migrations | table | postgres
(2 rows)

electric=# select * from items;
 id | value | inserted_at | updated_at
----+-------+-------------+------------
(0 rows)

electric=# insert into items (id) values (gen_random_uuid());
INSERT 0 1
electric=# \q
```

Now re-run the successful request and you'll get data:

```console
$ curl -s --header "Authorization: Bearer ${AUTH_TOKEN}" \
      "http://localhost:3000/v1/shape?table=items&offset=-1" | jq
[
  {
    "key": "\"public\".\"items\"/\"b702e58e-9364-4d54-9360-8dda20cb4405\"",
    "value": {
      "id": "b702e58e-9364-4d54-9360-8dda20cb4405",
      "value": null,
      "inserted_at": "2024-11-13 10:45:33",
      "updated_at": "2024-11-13 10:45:33"
    },
    "headers": {
      "operation": "insert",
      "relation": [
        "public",
        "items"
      ]
    },
    "offset": "0_0"
  }
]
```

> So far we've shown things working with Electric's lower-level [HTTP API](https://electric-sql.com/docs/api/http). You can also setup the [higher level clients](https://electric-sql.com/docs/api/clients/typescript) to use an auth token. See the [auth guide](https://electric-sql.com/docs/guides/auth) for more details.

Reviewer: i believe "higher level" should be hyphenated in "higher-level clients" because it is a multi-word adjective before a noun.

#### Proxy endpoint

4. the proxy validates the JWT auth token and verifies that the shape definition in the token matches the shape being requested; if so it reverse-proxies the request onto Electric
5. Electric then handles the request as normal
6. sending a response back *through the proxy* to the client
7. the client can then process the data and make additional requests using the same auth token (step 3); if the auth token expires or is rejected, the client starts again (step 1).

## How to run

There are three ways to run this example:

1. with the [API as both gatekeeper and proxy](#1-api-as-gatekeeper-and-proxy)
2. with the [API as gatekeeper and Caddy as the proxy](#2-caddy-as-proxy)
3. with the [API as gatekeeper and an edge function as the proxy](#3-edge-function-as-proxy)

It makes sense to run through these in order.

### Pre-reqs

You need [Docker Compose](https://docs.docker.com/compose/) and [curl](https://curl.se). We also (optionally) use [`psql`](https://www.postgresql.org/docs/current/app-psql.html) and pipe into [`jq`](https://jqlang.github.io/jq/) for JSON formatting.

> The instructions below all use the same [`./docker-compose.yaml`](./docker-compose.yaml) file in the root of this repo. With a different set of services and environment variables.

Reviewer 2: the dockerfile is not at the root of the repo, so this is confusing considering the hyperlink points to the right place

what about "the dockerfile in this folder"

### 1. API as gatekeeper and proxy

Build the local API image:

```shell
docker compose build api
```

Run `postgres`, `electric` and the `api` services:

```console
$ docker compose up postgres electric api
...
gatekeeper-api-1 | 10:22:20.951 [info] == Migrated 20241108150947 in 0.0s
gatekeeper-api-1 | 10:22:21.453 [info] Running ApiWeb.Endpoint with Bandit 1.5.7 at :::4000 (http)
gatekeeper-api-1 | 10:22:21.455 [info] Access ApiWeb.Endpoint at http://localhost:4000
```

In a new terminal, make a `POST` request to the gatekeeper endpoint:

```console
$ curl -sX POST "http://localhost:4000/gatekeeper/items" | jq
{
  "headers": {
    "Authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJKb2tlbiIsImV4cCI6MTczMTUwMjM2OSwiaWF0IjoxNzMxNDk1MTY5LCJpc3MiOiJKb2tlbiIsImp0aSI6IjMwM28zYmx0czN2aHYydXNiazAwMDJrMiIsIm5iZiI6MTczMTQ5NTE2OSwic2hhcGUiOnsibmFtZXNwYWNlIjpudWxsLCJ0YWJsZSI6Iml0ZW1zIiwid2hlcmUiOm51bGwsImNvbHVtbnMiOm51bGx9fQ.8UZehIWk1EDQ3dJ4ggCBNkx9vGudfrD9appqs8r6zRI"
  },

>   "url": "http://localhost:3000/proxy/v1/shape",

Reviewer 2: you have the port of the `/proxy` endpoint as 3000 here, but below you're making requests to `:4000/proxy/...`. 

For consistency the params returned from the gatekeeper auth request should work as config for the client, so I think the `url` here should be `http://localhost:4000/proxy/v1/shape`

and in fact in the next para you say the url is `:4000` so this is a typo

  },
  "url": "http://localhost:8080/v1/shape",
  "table": "items"
}
```

Copy the auth token and set it to an env var:

```shell
export AUTH_TOKEN="<token>"
```

An unauthorised request to Caddy will get a 401:

```console
$ curl -sv "http://localhost:8080/v1/shape?table=items&offset=-1"
...
< HTTP/1.1 401 Unauthorized
< Server: Caddy
...
```

An authorised request for the correct shape will succeed:

```console
$ curl -sv --header "Authorization: Bearer ${AUTH_TOKEN}" \
      "http://localhost:8080/v1/shape?table=items&offset=-1"
...
< HTTP/1.1 200 OK
...
```

Caddy validates the shape request against the shape definition signed into the auth token. So an authorised request *for the wrong shape* will fail:

```console
$ curl -sv --header "Authorization: Bearer ${AUTH_TOKEN}" \
      "http://localhost:8080/v1/shape?table=items&offset=-1&where=true"
...
< HTTP/1.1 403 Forbidden
...
```

> Take a look at the [`./caddy/Caddyfile`](./caddy/Caddyfile) for more details.

Reviewer 2: Impressive Caddyfile foo 🥇
